### DISCUSSION TITLE
Proposal: retaining moderator role with a "hashed" room

### DISCUSSION THREAD

**Discussion by edocod1** [edumeet/edumeet#852] [Ideas]
Hello,
we are having some issues with MM and people that use it. Basically, it is hard to assign user roles and if the creator of the room loses connection their status is lost.

To fix this, I think a simple method with hashed rooms could be implemented. Instead of creating a room with an arbitrary name, the name would be an hash of another string sent as a GET parameter together with the room.

Basically, a hashed room would look like this:
`https://[server_name]/[hash_of_passphrase]?[passphrase]`

`https://host/42a00eaa9468b08adc80d9dfb877a6d30c3ed64eea293e945228669ee3de4f1e?mypassword` (i used sha3-256 here, but any could be used)

Users that join using the hashed url (with the ?[password] parameter) would be given admin rights, while all the other users would be normal guests.

Do you think it could be implemented? We are having a ton of users asking for this feature, since many are getting locked out if they lock a room and lose connection, or if links they've shared in calendars are joined by people earlier than them.

**Comment by giuliolunati** [comment]
Would be good!
But I think "normal" names should keep normal behaviour.
So:
/a-normal-room => current behaviour

/hash?passphrase => I'm a boss, I can mute others, lock the room, kick off people../

/hash* => I'm unprivileged, I can't lock the room

\* hash should be recognized by being a fixed number of hex

**Comment by edocod1** [comment]
Of course normal names will keep the normal behaviour. However, people that for some reason use hashes to create rooms (perhaps automated systems to have unique rooms?) not aware of this will find themselves as guests for no apparent reason.

**Comment by giuliolunati** [comment]
Maybe we can use a different pattern, e.g.
/@hash?passphrase

**Comment by cap19** [comment]
Have you looked at the new user roles and BYPASS_ROOM_LOCK features for certain roles in the develop stream?

It might solve a few of the issues you are running into.

**Comment by misi** [comment]
@edocod1  use Open ID connect to get persistent info about user and use user map function to upgrade user based on it to moderator privilege.

**Comment by misi** [comment]
You can create own OIDC OP that provides user info and acts as Identity Provider. 
There are so many open implementation of it. Some are also certified see:
 https://openid.net/certification/
You could also use any open source identity gw platform. e.g. keycloak, satosa, or any similar
with it you can get connection to SAML LDAP, sql User db, or more to ger user info form other sources.

**Comment by edocod1** [comment]
Hello misi,
yes, I am aware of that. But that would mean storing user data, and other than the various complications due to GDPR, that'd defy the immediateness of the service and add extra steps users have to go through.

**Comment by misi** [comment]
@edocod1 development branch address the reconnect issue AFAIR. 

Yes we can consider something similar auth mechanism you proposed with room names and moderator password, but I still think we should leave AAI for AAI standard protocols.

Unfortunately I am too busy actually to implement it right now.
Please send a Pull Request and I am happy to merge it, if not I will leave this issue open as new feature request.

**Comment by tapionx** [comment]
OIDC is useful for sure, but it adds a layer of complexity that must be avoided in some use cases:

the setup of an identity provider is complex and error prone
storing usernames and passwords in a database raises privacy concerns and it makes harder to comply with GDPR
more action for the users (sign-in) means more friction in the UX
iorestoacasa.work is offering a videocalls for free, without authentication.
The lack of moderation features is affecting the experience for our community, because you want to be able to kick intruders or mute everyone if the room is large.

I propose a simple and effective mechanism to enable moderation roles without authentication.

When a room is created, the user is provided with a moderation key computed this way:
sha256( $room_name + $cookie_secret + $room_creation_timestamp)

Users can join the room with this moderation key as a GET parameter

https://example.com/room-name?moderation_key=XXXXXXX

and they will gain the MODERATOR role.

**Comment by edocod1** [comment]
@tapionx Hmm, i think the problem with that solution would be that it'd add the problem that now random people would be able to gain access to arbitrary rooms anyway (imagine that i announce a webinar in one month, and one month later somebody already joined and gained admin) and now we'd have to store all the room names and eventually the creation timestamp for an indefinite time.

I'd add instead a mechanism of public/private rooms.

During room creation the user would be asked if they want to create a public or private room. Public rooms are the same as always (to not disrupt users already familiar with the usual method), but private rooms have a completely different endpoint, and the name of them is a public key derived from a randomly generated private key (that is shown to the user when they generate the room).

This way you wouldn't have to store any metadata, and you'd prevent room stealing (you cannot regenerate the same room twice), having to store any kind of data (the keypair is verified on the fly, no state involved), cookies/credentials expiration (keypairs are valid forever) and you'd even be able to share the keypair to other people you want to give admin privileges to.
